x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/RequestExecutorService.java

```diff
@@ -49,51 +48,105 @@
+    private final AdjustableCapacityBlockingQueue<AbstractRunnable> queue;
     private final AtomicBoolean running = new AtomicBoolean(true);
     private final CountDownLatch terminationLatch = new CountDownLatch(1);
     private final HttpClientContext httpContext;
     private final HttpClient httpClient;
     private final ThreadPool threadPool;
     private final CountDownLatch startupLatch;
+    private final BlockingQueue<Runnable> controlQueue = new LinkedBlockingQueue<>();
 
-    @SuppressForbidden(reason = "wraps a queue and handles errors appropriately")
-    HttpRequestExecutorService(String serviceName, HttpClient httpClient, ThreadPool threadPool, @Nullable CountDownLatch startupLatch) {
-        this(serviceName, httpClient, threadPool, new LinkedBlockingQueue<>(), startupLatch);
-    }
-
-    @SuppressForbidden(reason = "wraps a queue and handles errors appropriately")
-    HttpRequestExecutorService(
+    RequestExecutorService(
         String serviceName,
         HttpClient httpClient,
         ThreadPool threadPool,
-        int capacity,
-        @Nullable CountDownLatch startupLatch
+        @Nullable CountDownLatch startupLatch,
+        RequestExecutorServiceSettings settings
     ) {
-        this(serviceName, httpClient, threadPool, new LinkedBlockingQueue<>(capacity), startupLatch);
+        this(serviceName, httpClient, threadPool, RequestExecutorService::buildQueue, startupLatch, settings);
+    }
+
+    private static BlockingQueue<AbstractRunnable> buildQueue(int capacity) {
+        BlockingQueue<AbstractRunnable> queue;
+        if (capacity <= 0) {
+            queue = new LinkedBlockingQueue<>();
+        } else {
+            queue = new LinkedBlockingQueue<>(capacity);
+        }
+
+        return queue;
     }
 
     /**
      * This constructor should only be used directly for testing.
      */
-    @SuppressForbidden(reason = "wraps a queue and handles errors appropriately")
-    HttpRequestExecutorService(
+    RequestExecutorService(
         String serviceName,
         HttpClient httpClient,
         ThreadPool threadPool,
-        BlockingQueue<HttpTask> queue,
-        @Nullable CountDownLatch startupLatch
+        Function<Integer, BlockingQueue<AbstractRunnable>> createQueue,
+        @Nullable CountDownLatch startupLatch,
+        RequestExecutorServiceSettings settings
     ) {
         this.serviceName = Objects.requireNonNull(serviceName);
         this.httpClient = Objects.requireNonNull(httpClient);
         this.threadPool = Objects.requireNonNull(threadPool);
         this.httpContext = HttpClientContext.create();
-        this.queue = queue;
+        this.queue = new AdjustableCapacityBlockingQueue<>(settings.getQueueCapacity(), createQueue);
         this.startupLatch = startupLatch;
+
+        Objects.requireNonNull(settings);
+        settings.registerQueueCapacityCallback(this::onCapacityChange);
+    }
+
+    private void onCapacityChange(int capacity) {
```

Comment:
This will be called when the `ClusterService` receives a change event for the queue capacity setting. That calling logic is encapsulated within the `RequestExecutorServiceSettings`